fireplace/dsl/random_picker.py

```diff
@@ -12,48 +12,128 @@ def __init__(self, *args, **filters):
 
 	def __mul__(self, other):
 		ret = self.__class__(*self.args, **self.filters)
+		ret._cards = self._cards
+		ret._source = self._source
 		ret.count = other
 		return ret
 
-	@property
-	def cards(self):
-		if self._cards is None:
-			self._cards = self._filter_cards(self.filters)
-		return self._cards
-
 	def _filter_cards(self, filters):
 		from .. import cards
 		return cards.filter(**filters)
 
-	def get_cards(self, source):
-		filters = self.filters.copy()
+	def _get_lazy_filtered_cards(self, source, filters):
+		newFilters = filters.copy()
 		# Iterate through the filters, evaluating the LazyValues as we go
-		for k, v in filters.items():
+		for k, v in newFilters.items():
 			if isinstance(v, LazyValue):
-				filters[k] = v.evaluate(source)
-		return self._filter_cards(filters)
-
-	def evaluate(self, source) -> str:
-		if self.lazy_filters:
-			# If the card has lazy filters, we need to evaluate them
-			cards = self.get_cards(source)
-		else:
-			cards = self.cards
+				newFilters[k] = v.evaluate(source)
+		return self._filter_cards(newFilters)
+
+	def find_cards(self, source=None, **filters):
+		if len(filters)==0:
+			filters = self.filters
+
+		if source is None:
+			source = self._source
+
+		if self._cards is None or filters != self.filters or source != self._source:
+			lazy_filters = False
+			for v in filters.values():
+				if isinstance(v, LazyValue):
+					lazy_filters = True
+					break
+
+			if lazy_filters:
+				# If the card has lazy filters, we need to evaluate them
+				self._cards = self._get_lazy_filtered_cards(source, filters)
+			else:
+				self._cards = self._filter_cards(filters)
+
+		return self._cards
+
+	def evaluate(self, source, cards=None) -> str:
+		cards = cards or self.find_cards(source)
 		ret = random.sample(cards, self.count)
 		return [source.controller.card(card, source=source) for card in ret]
 
 
+# get random card(s) matching the filter in the constructor plus one of several weighted filters
+class RandomCardWeighted(RandomCardPicker):
+	def __init__(self, *args, **filters):
+		self._weights = []
+		self._weightedfilters = []
+
+		super().__init__(*args, **filters)
+
+	# add a filter set
+	def add(self, weight, **filters):
+		self._weights.append(weight)
+		self._weightedfilters.append(filters)
+		return self
+
+	# select number of cards to fetch (overrides base __mul__ because creating a new object will cause the derived members to be lost)
+	def __mul__(self, other):
+		ret = super().__mul__(other)
+		ret._weights = self._weights
+		ret._weightedfilters = self._weightedfilters
+		return ret
+
+	# This picks from a single combined card pool without replacement, weighting each filtered set of cards against the total
+	def evaluate(self, source):
+
+		# add the global filters to each set of filters
+		self._weightedfilters[:] = [{ **x, **self.filters } for x in self._weightedfilters]
+
+		# get all the cards matching each set of filters and create the weight line
```

Comment:
This seems a bit complex.

Can you factor out a helper function, e.g. weighted_choice(choices, weights) ? That would simplify the logic significantly, since you have the choices and the weights readily available.